pandas-stubs/core/arrays/sparse/array.pyi

```diff
@@ -1,62 +1,97 @@
+from collections.abc import (
+    Callable,
+    Hashable,
+)
 from enum import Enum
+import sys
 from typing import (
     Any,
+    Literal,
+    Protocol,
     final,
     overload,
+    type_check_only,
 )
 
+import numpy as np
 from pandas.core.arrays import ExtensionArray
+from pandas.core.series import Series
 from typing_extensions import Self
 
+from pandas._libs.sparse import SparseIndex
 from pandas._typing import (
+    AnyArrayLike,
+    Axis,
+    AxisInt,
     NpDtype,
+    Scalar,
     ScalarIndexer,
     SequenceIndexer,
     np_1darray,
+    np_ndarray,
+    np_ndarray_int,
 )
 
+from pandas.core.dtypes.dtypes import SparseDtype
+
+@type_check_only
+class _SparseMatrixLike(Protocol):
+    @property
+    def shape(self, /) -> tuple[int, int]: ...
+
 @final
 class ellipsis(Enum):
     Ellipsis = "..."
 
 class SparseArray(ExtensionArray):
-    def __init__(
-        self,
-        data,
-        sparse_index=...,
-        fill_value=...,
-        kind: str = ...,
-        dtype=...,
-        copy: bool = ...,
-    ) -> None: ...
+    if sys.version_info >= (3, 11):
+        def __new__(
+            cls,
+            data: AnyArrayLike | Scalar,
+            sparse_index: SparseIndex | None = None,
+            fill_value: Scalar | None = None,
+            kind: str = "integer",
+            dtype: np.dtype | SparseDtype | None = ...,
+            copy: bool = ...,
+        ) -> Self: ...
+    else:
+        def __new__(
+            cls,
+            data: AnyArrayLike | Scalar,
+            sparse_index: SparseIndex | None = None,
+            fill_value: Scalar | None = None,
+            kind: str = "integer",
+            dtype: np.dtype[Any] | SparseDtype | None = ...,
+            copy: bool = ...,
+        ) -> Self: ...
+
     @classmethod
-    def from_spmatrix(cls, data) -> Self: ...
+    def from_spmatrix(cls, data: _SparseMatrixLike) -> Self: ...
     def __array__(
         self, dtype: NpDtype | None = None, copy: bool | None = None
     ) -> np_1darray: ...
-    def __setitem__(self, key, value) -> None: ...
```

Comment:
https://github.com/pandas-dev/pandas/blob/13fc5105ca42f3ba5b986d9ae7ec6076b3652136/pandas/core/arrays/sparse/array.py#L605-L612

Deleting for now, seems like it raises an error in any case, same in boolean